// restify-plugins/lib/plugins/auditFormatter.js
// Copyright 2017 restify.  All rights reserved.
// Copyright 2012 Mark Cavage Inc. All rights reserved.

'use strict';

var assert = require('assert-plus');
var bunyan = require('bunyan');
var HttpError = require('restify-errors').HttpError;


///--- API


/**
 * @param {object} opts The options object.
 * @param {object} opts.log The logger.
 * @param {string} [opts.event] The event from the server which initiates the
 * audit log.
 * @param {object} [opts.server] The restify server, used to emit the audit log
 * object programttically
 * @param {boolean} [opts.print=true] Whether to print the log via the logger.
 *
 * @public
 * @function afterLogger
 * @returns {Function}
 * @emits audit when an audit log has been generated
 */
function afterLogger(opts) {
    assert.object(opts, 'opts');
    assert.object(opts.log, 'opts.log');
    assert.optionalString(opts.event, 'opts.event');
    assert.optionalObject(opts.server, 'opts.server');
    assert.optionalBool(opts.print, 'opts.print');

    var server = opts.server;
    var print = opts.print;

    if (typeof print === 'undefined') {
        print = true;
    }

    // setup bunyan logger
    var errSerializer = bunyan.stdSerializers.err;

    if (opts.log.serializers && opts.log.serializers.err) {
        errSerializer = opts.log.serializers.err;
    }

    var log = opts.log.child({
        audit: true,
        component: opts.event,
        serializers: {
            err: errSerializer,
            req: function auditRequestSerializer(req) {
                if (!req) {
                    return (false);
                }

                var timers = {};
                (req.timers || []).forEach(function (time) {
                    var t = time.time;
                    var _t = Math.floor((1000000 * t[0]) + (t[1] / 1000));
                    timers[time.name] = _t;
                });
                return ({
                    // account for native and queryParser plugin usage
                    query: (typeof req.query === 'function') ?
                            req.query() : req.query,
                    method: req.method,
                    url: req.url,
                    headers: req.headers,
                    httpVersion: req.httpVersion,
                    trailers: req.trailers,
                    version: req.version(),
                    body: opts.body === true ? req.body : undefined,
                    timers: timers,
                    connectionState: (req.connectionState &&
                                     req.connectionState())
                });
            },
            res: function auditResponseSerializer(res) {
                if (!res) {
                    return (false);
                }

                var body;

                if (opts.body === true) {
                    if (res._body instanceof HttpError) {
                        body = res._body.body;
                    } else {
                        body = res._body;
                    }
                }

                return ({
                    statusCode: res.statusCode,
                    headers: res._headers,
                    trailer: res._trailer || false,
                    body: body
                });
            }
        }
    });

    function audit(req, res, route, err) {
        var latency = res.get('Response-Time');

        if (typeof (latency) !== 'number') {
            latency = Date.now() - req._time;
        }

        var obj = {
            remoteAddress: req.connection.remoteAddress,
            remotePort: req.connection.remotePort,
            req_id: req.getId(),
            req: req,
            res: res,
            err: err,
            latency: latency,
            secure: req.secure,
            _audit: true,
            event: opts.event
        };

        if (print) {
            log.info(obj, 'handled: %d', res.statusCode);
        }

        if (server) {
            server.emit('afterlog', obj);
        }

        return (true);
    }

    return (audit);
}


///-- Exports


module.exports = afterLogger;
